File: src/mcp/src/kifu.ts
import { GameState, MoveRecord, PieceType, PositionSnapshot } from './types.js';
import { applyMoveRecord, cloneState, coordToKifu, createInitialState, HashCalc, kifuToCoord, moveToJapanese, stateToSfenLike } from './shogi.js';

export const KifuCodec = {
  coordToStr: coordToKifu,
  strToCoord: kifuToCoord,

  pieceToChar(type: string): string {
    switch (type) {
      case 'OU':
        return 'O';
      case 'KIN':
        return 'G';
      case 'KAKU':
      case 'UMA':
        return 'K';
      default:
        return '?';
    }
  },

  charToPiece(char: string): PieceType {
    switch (char) {
      case 'O':
        return 'OU';
      case 'G':
        return 'KIN';
      case 'K':
        return 'KAKU';
      default:
        throw new Error(`不正な駒文字です: ${char}`);
    }
  },

  encodeMove(move: MoveRecord): string {
    if (move.type === 'board') {
      return `${coordToKifu(move.fromX, move.fromY)}${coordToKifu(move.toX, move.toY)}${move.promote ? '+' : ''}`;
    }
    return `${this.pieceToChar(move.piece)}${coordToKifu(move.toX, move.toY)}`;
  },

  decodeMove(value: string): MoveRecord {
    if (/^[A-Z]/.test(value)) {
      const piece = this.charToPiece(value[0] ?? '');
      const to = kifuToCoord(value.slice(1, 3));
      return { type: 'hand', toX: to.x, toY: to.y, piece, promote: false };
    }
    const from = kifuToCoord(value.slice(0, 2));
    const to = kifuToCoord(value.slice(2, 4));
    return { type: 'board', fromX: from.x, fromY: from.y, toX: to.x, toY: to.y, promote: value.includes('+') };
  },

  encode(moves: MoveRecord[]): string {
    return moves.map((move) => this.encodeMove(move)).join('-');
  },

  decode(value: string): MoveRecord[] {
    const kifu = extractKifuString(value);
    if (!kifu) return [];
    return kifu.split('-').filter(Boolean).map((part) => this.decodeMove(part));
  },

  moveToJapanese,
};

export function extractKifuString(input: string): string {
  const value = input.trim();
  if (!value) return '';

  try {
    if (value.includes('://')) {
      return new URL(value).searchParams.get('m') ?? '';
    }
  } catch {
    throw new Error(`棋譜URLを解析できません: ${input}`);
  }

  const query = value.startsWith('?') ? value.slice(1) : value;
  if (query.startsWith('m=')) return new URLSearchParams(query).get('m') ?? '';
  return value;
}

export function replayKifu(input: string, options: { moveLimit?: number } = {}) {
  const records = KifuCodec.decode(input);
  const limit = options.moveLimit ?? records.length;
  let state: GameState = createInitialState();
  const positions: PositionSnapshot[] = [snapshot(0, state)];
  const moves: Array<{ ply: number; record: MoveRecord; japanese: string; hash: string }> = [];

  for (const [index, record] of records.slice(0, limit).entries()) {
    const applied = applyMoveRecord(state, record);
    state = applied.state;
    const position = snapshot(index + 1, state);
    positions.push(position);
    moves.push({ ply: index + 1, record: applied.record, japanese: applied.japanese, hash: position.hash });
  }

  return {
    input: extractKifuString(input),
    moves,
    positions,
    finalPosition: positions[positions.length - 1],
  };
}

function snapshot(ply: number, state: GameState): PositionSnapshot {
  const cloned = cloneState(state);
  return {
    ply,
    turn: cloned.turn,
    hash: HashCalc.hashToString(HashCalc.calcHash(cloned)),
    sfenLike: stateToSfenLike(cloned),
    state: cloned,
  };
}

File: src/mcp-go/kifu.go
package main

import (
	"fmt"
	"net/url"
	"strconv"
	"strings"
)

func pieceToChar(pt PieceType) string {
	switch pt {
	case OU:
		return "O"
	case KIN:
		return "G"
	case KAKU, UMA:
		return "K"
	default:
		return "?"
	}
}

func charToPiece(char string) (PieceType, error) {
	switch char {
	case "O":
		return OU, nil
	case "G":
		return KIN, nil
	case "K":
		return KAKU, nil
	default:
		return "", fmt.Errorf("不正な駒文字です: %s", char)
	}
}

func encodeMove(record MoveRecord) string {
	switch r := record.(type) {
	case BoardMoveRecord:
		promoteStr := ""
		if r.Promote {
			promoteStr = "+"
		}
		return fmt.Sprintf("%s%s%s", coordToKifu(r.FromX, r.FromY), coordToKifu(r.ToX, r.ToY), promoteStr)
	case HandMoveRecord:
		return fmt.Sprintf("%s%s", pieceToChar(r.Piece), coordToKifu(r.ToX, r.ToY))
	}
	return ""
}

func decodeMove(value string) (MoveRecord, error) {
	if len(value) == 0 {
		return nil, fmt.Errorf("空の文字列です")
	}
	if value[0] >= 'A' && value[0] <= 'Z' {
		piece, err := charToPiece(string(value[0]))
		if err != nil {
			return nil, err
		}
		toX, toY, err := kifuToCoord(value[1:3])
		if err != nil {
			return nil, err
		}
		return HandMoveRecord{Type: "hand", ToX: toX, ToY: toY, Piece: piece, Promote: false}, nil
	}
	fromX, fromY, err := kifuToCoord(value[0:2])
	if err != nil {
		return nil, err
	}
	toX, toY, err := kifuToCoord(value[2:4])
	if err != nil {
		return nil, err
	}
	return BoardMoveRecord{
		Type:    "board",
		FromX:   fromX,
		FromY:   fromY,
		ToX:     toX,
		ToY:     toY,
		Promote: strings.Contains(value, "+"),
	}, nil
}

func encodeKifu(moves []MoveRecord) string {
	parts := make([]string, len(moves))
	for i, move := range moves {
		parts[i] = encodeMove(move)
	}
	return strings.Join(parts, "-")
}

func decodeKifu(value string) ([]MoveRecord, error) {
	kifu := extractKifuString(value)
	if kifu == "" {
		return nil, nil
	}
	parts := strings.Split(kifu, "-")
	var records []MoveRecord
	for _, part := range parts {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		record, err := decodeMove(part)
		if err != nil {
			return nil, err
		}
		records = append(records, record)
	}
	return records, nil
}

func extractKifuString(input string) string {
	value := strings.TrimSpace(input)
	if value == "" {
		return ""
	}
	if strings.Contains(value, "://") {
		u, err := url.Parse(value)
		if err != nil {
			return ""
		}
		return u.Query().Get("m")
	}
	query := value
	if strings.HasPrefix(query, "?") {
		query = query[1:]
	}
	if strings.HasPrefix(query, "m=") {
		params, err := url.ParseQuery(query)
		if err != nil {
			return ""
		}
		return params.Get("m")
	}
	return value
}

type ReplayResult struct {
	Input         string
	Moves         []ReplayMove
	Positions     []PositionSnapshot
	FinalPosition *PositionSnapshot
}

type ReplayMove struct {
	Ply      int
	Record   MoveRecord
	Japanese string
	Hash     string
}

func replayKifu(input string, moveLimit *int) (*ReplayResult, error) {
	records, err := decodeKifu(input)
	if err != nil {
		return nil, fmt.Errorf("棋譜のデコードに失敗しました: %w", err)
	}
	limit := len(records)
	if moveLimit != nil && *moveLimit < limit {
		limit = *moveLimit
	}

	state := createInitialState()
	positions := []PositionSnapshot{snapshot(0, state)}
	var moves []ReplayMove

	for i := 0; i < limit; i++ {
		record := records[i]
		newState, enriched, japanese, err := applyMoveRecord(state, record)
		if err != nil {
			return nil, fmt.Errorf("手 %d の適用に失敗しました: %w", i+1, err)
		}
		state = newState
		pos := snapshot(i+1, state)
		positions = append(positions, pos)
		moves = append(moves, ReplayMove{
			Ply:      i + 1,
			Record:   enriched,
			Japanese: japanese,
			Hash:     pos.Hash,
		})
	}

	result := &ReplayResult{
		Input:     extractKifuString(input),
		Moves:     moves,
		Positions: positions,
	}
	if len(positions) > 0 {
		final := positions[len(positions)-1]
		result.FinalPosition = &final
	}
	return result, nil
}

func snapshot(ply int, state GameState) PositionSnapshot {
	cloned := cloneState(state)
	return PositionSnapshot{
		Ply:      ply,
		Turn:     cloned.Turn,
		Hash:     HashCalc.HashToString(HashCalc.CalcHash(cloned)),
		SFENLike: stateToSFENLike(cloned),
		State:    cloned,
	}
}

func resolvePosition(input PositionInput) (*PositionSnapshot, error) {
	if input.State != nil {
		ply := 0
		if input.Ply != nil {
			ply = *input.Ply
		}
		pos := snapshot(ply, *input.State)
		return &pos, nil
	}

	kifu := ""
	if input.Kifu != nil {
		kifu = *input.Kifu
	} else if input.Input != nil {
		kifu = *input.Input
	}

	result, err := replayKifu(kifu, nil)
	if err != nil {
		return nil, err
	}

	ply := len(result.Positions) - 1
	if input.Ply != nil {
		ply = *input.Ply
	}
	if ply < 0 || ply >= len(result.Positions) {
		return nil, fmt.Errorf("指定手数の局面がありません: %d", ply)
	}
	pos := result.Positions[ply]
	return &pos, nil
}

// PositionInput はツールの入力パラメータ
type PositionInput struct {
	Input *string    `json:"input,omitempty"`
	Kifu  *string    `json:"kifu,omitempty"`
	Ply   *int       `json:"ply,omitempty"`
	State *GameState `json:"state,omitempty"`
}

// DecodeKifuInput はdecode_kifu ツールの入力
type DecodeKifuInput struct {
	Input     string `json:"input"`
	MoveLimit *int   `json:"moveLimit,omitempty"`
}

// BestLineInput はget_best_line ツールの入力
type BestLineInput struct {
	PositionInput
	MaxPlies *int `json:"maxPlies,omitempty"`
}

// parseInt のヘルパー
func mustParseInt(s string) (int, error) {
	return strconv.Atoi(s)
}